### PR TITLE
Add GloParT inference facility

### PR DESCRIPTION
**Pull request by colizz**
#### PR description:

This PR adds the GloParT 3 model's inference facility into CMSSW, prepared for NanoAODv15, replacing #45830. 
Model performance details are provided in these [[slides]](https://cernbox.cern.ch/files/spaces/eos/user/c/coli/slides/24.10.18_GloParT%203%20review.pdf) (accessible via CMS).

Global Particle Transformer (GloParT) 3 is an inclusive tagging model for AK8 jets that covers the entire phase space and enables resonance mass regression for each jet class. It functions as both a global tagger and a mass regression model for AK8 jets and can also be utilized as a pre-trained model. The hidden layer neurons (with dimension 256) are stored in MiniAOD, providing the capability to resume all output scores. Further details can be found in the slides.

In summary, the integration overview (the new AK8 jet features added to MiniAOD and NanoAOD) is illustrated below.

glopart-integration

Please test this PR with https://github.com/cms-data/RecoBTag-Combined/pull/62.

#### PR validation:

 - The PR passes the unit test `test_globalpart_cfg.py` and is also validated under the AODSIM->MINIAOD->NANOAOD workflow.
 - Profiling performance in cmssw is delivered and shown in the slides above (see pp.89-93).
 - The output scores produced via ONNX in the cmssw workflow are validated to match those in the PyTorch workflow.

### CODE REVIEW

**Comment by jfernan2** [comment]
enable nano

**Comment by colizz** [comment]
I think it's caused by not testing with https://github.com/cms-data/RecoBTag-Combined/pull/62
Could you help test again including this test parameter? Thanks

**Comment by hqucms** [comment]
test parameters:
 - pull_request = https://github.com/cms-data/RecoBTag-Combined/pull/62
 - enable = nano,profiling

**Comment by hqucms** [comment]
please test

File: RecoBTag/FeatureTools/plugins/GlobalParticleTransformerAK8TagInfoProducer.cc [hunk 1]
```diff
@@ -0,0 +1,742 @@
...
+    } else {
+      npfPtrs.push_back(cand);
+    }
+  }
+  // lost tracks: fill to cpfcands
+  for (size_t i = 0; i < lts_->size(); ++i) {
+    auto cand = lts_->ptrAt(i);
+    if (reco::deltaR(*cand, jet) < jet_radius_) {
+      cpfPtrs.push_back(cand);
+      isLostTrackMap[cand.key()] = true;
```
Comment (hqucms) [inline RecoBTag/FeatureTools/plugins/GlobalParticleTransformerAK8TagInfoProducer.cc:402]: @colizz Is `cand.key()` unique for lost tracks and packed candidates? Or the `key()` is just the index in each collection (then different types of objects can have the same keys)? Could you please check?

**Comment by nurfikri89** [comment]
@colizz Thanks for making the PR. I think you need to add the GloParT3 branches in `nanoDQM_cfi.py` also.

File: RecoBTag/FeatureTools/plugins/GlobalParticleTransformerAK8TagInfoProducer.cc (hunk 1 above)
Comment (colizz) [reply RecoBTag/FeatureTools/plugins/GlobalParticleTransformerAK8TagInfoProducer.cc:402]: hi @hqucms , they are unique keys for `reco::CandidatePtr` hence should be different for each lost track and packed candidates. 
<details>
  <summary>Here is my check:</summary>
```
cpf key: 37
cpf key: 39
cpf key: 42
cpf key: 46
cpf key: 249
cpf key: 1004
cpf key: 1012
cpf key: 1013
cpf key: 1014
cpf key: 1016
cpf key: 1017
cpf key: 1018
cpf key: 1019
cpf key: 1020
cpf key: 1024
cpf key: 1025
cpf key: 1028
cpf key: 1029
cpf key: 1030
cpf key: 1032
cpf key: 1036
cpf key: 1039
cpf key: 1040
cpf key: 1041
cpf key: 1042
cpf key: 1044
cpf key: 1051
cpf key: 1052
cpf key: 1304
cpf key: 44
cpf key: 50
cpf key: 51
cpf key: 1022
cpf key: 1023
cpf key: 1026
cpf key: 1027
cpf key: 1031
cpf key: 1033
cpf key: 1034
cpf key: 1035
cpf key: 1037
cpf key: 1038
cpf key: 1043
cpf key: 1045
cpf key: 1046
cpf key: 1049
cpf key: 1056
cpf key: 1069
cpf key: 1219
lts key: 16
lts key: 17
lts key: 47
=== end of jets ===
cpf key: 47
cpf key: 56
cpf key: 978
cpf key: 1001
cpf key: 1003
cpf key: 1005
cpf key: 1053
cpf key: 1055
cpf key: 1057
cpf key: 1058
cpf key: 1059
cpf key: 1060
cpf key: 1061
cpf key: 1062
cpf key: 1063
cpf key: 1064
cpf key: 1065
cpf key: 1066
cpf key: 1068
cpf key: 1134
cpf key: 57
cpf key: 998
cpf key: 1067
cpf key: 1071
cpf key: 1072
cpf key: 1073
cpf key: 1074
cpf key: 1075
cpf key: 1076
cpf key: 1048
cpf key: 1054
cpf key: 1070
cpf key: 387
lts key: 7
lts key: 10
lts key: 12
lts key: 13
lts key: 14
lts key: 19
lts key: 49
=== end of jets ===
```
</details>

Comment (colizz) [reply RecoBTag/FeatureTools/plugins/GlobalParticleTransformerAK8TagInfoProducer.cc:402]: thanks a lot @hqucms , this is very nice catch. There is an overlap between "keys" in PF candidates and lost tracks, with a prob of ~0.5% (though small, it indicates a logical error in the code).
We should point out that in training GloParT, we used the same code to generate training samples, which means that a small portion of charged PF candidates were mistakenly labelled as lost tracks. For now, it appears that this error is consistent across both, allowing our trained GloParT model to perform the same in CMSSW inference.

It seems our best option is to retain this code to preserve GloParT's performance. Perhaps we could add a comment here for clarity. What do you think?

Comment (hqucms) [reply RecoBTag/FeatureTools/plugins/GlobalParticleTransformerAK8TagInfoProducer.cc:402]: @colizz Given the small possibility of the overlap, I would expect fixing this should have a rather negligible effect on the performance? If that is the indeed case I think it's better to fix it now already.

Comment (colizz) [reply RecoBTag/FeatureTools/plugins/GlobalParticleTransformerAK8TagInfoProducer.cc:402]: ok, we will check whether the effect is negligible if we change the code here.

**Comment by hqucms** [comment]
please abort

**Comment by hqucms** [comment]
please test

**Comment by hqucms** [comment]
please abort

**Comment by hqucms** [comment]
please test

File: RecoBTag/FeatureTools/plugins/GlobalParticleTransformerAK8TagInfoProducer.cc (hunk 1 above)
Comment (colizz) [reply RecoBTag/FeatureTools/plugins/GlobalParticleTransformerAK8TagInfoProducer.cc]: hi @hqucms , the code has been fixed to use `CandidatePtr` itself as the map key: https://github.com/cms-sw/cmssw/pull/46523/commits/b813fb6bcb21869258145de96d49fa20136bc613
Additionally, `puppi_wgt_cache` had a similar issue; however, testing revealed that it does not produce the same errors as `isLostTrackMap`. In this updated code, we validated the scores and found that in <1% of cases, some charged PF cands are marked as `isLostTrack=True`, but this has only a minimal impact on the GloParT output scores. We also examined the ROC curve on a large sample set, and found no change to the ROC performance. Thus we conclude that this modification has a negligible impact on GloParT performance

**Comment by colizz** [comment]
> @colizz Thanks for making the PR. I think you need to add the GloParT3 branches in `nanoDQM_cfi.py` also.

@nurfikri89 thanks! I updated `nanoDQM_cfi.py` to include the new branches: https://github.com/cms-sw/cmssw/pull/46523/commits/243edc69092488ae4ed876479364b2cefa979658

File: RecoBTag/FeatureTools/plugins/GlobalParticleTransformerAK8TagInfoProducer.cc (hunk 1 above)
Comment (hqucms) [reply RecoBTag/FeatureTools/plugins/GlobalParticleTransformerAK8TagInfoProducer.cc]: Thanks a lot @colizz for the fast turnaround! The fix looks good to me.

**Comment by hqucms** [comment]
please abort

**Comment by hqucms** [comment]
please test

**Comment by nurfikri89** [comment]
> > @colizz Thanks for making the PR. I think you need to add the GloParT3 branches in `nanoDQM_cfi.py` also.
> 
> @nurfikri89 thanks! I updated `nanoDQM_cfi.py` to include the new branches: https://github.com/cms-sw/cmssw/pull/46523/commits/243edc69092488ae4ed876479364b2cefa979658

Brilliant. Looks good. Thanks!

**Comment by hqucms** [comment]
@smuzaffar There seems to be some trouble in getting the tests to run through for this PR. Could you please have a look?

**Comment by smuzaffar** [comment]
@hqucms , the profiling jobs for this PR are still in [queue](https://cmssdt.cern.ch/jenkins/job/ib-run-pr-profiling/) ( waiting for resources to be available)

**Comment by hqucms** [comment]
> @hqucms , the profiling jobs for this PR are still in [queue](https://cmssdt.cern.ch/jenkins/job/ib-run-pr-profiling/) ( waiting for resources to be available)

I see. Thanks a lot!

**Comment by hqucms** [comment]
test parameters:
 - pull_request = https://github.com/cms-data/RecoBTag-Combined/pull/62
 - enable = nano

**Comment by hqucms** [comment]
please test

**Comment by hqucms** [comment]
+1

**Comment by jfernan2** [comment]
+1

**Comment by mandrenguyen** [comment]
+1

**Comment by smuzaffar** [comment]
@colizz , there is a request to backport [RecoBTag-Combined tag V01-25-00](https://github.com/cms-sw/cmsdist/pull/9505) to 14.1.X. Do we need a backport of this PR to also go with https://github.com/cms-sw/cmsdist/pull/9505

**Comment by hqucms** [comment]
@smuzaffar No, the changes introduced in V01-25-00 (i.e., https://github.com/cms-data/RecoBTag-Combined/pull/63) is unrelated to this PR.

**Comment by smuzaffar** [comment]
what about changes in https://github.com/cms-data/RecoBTag-Combined/pull/62 ? https://github.com/cms-sw/cmsdist/pull/9505 include changes from both  https://github.com/cms-data/RecoBTag-Combined/pull/62 and  https://github.com/cms-data/RecoBTag-Combined/pull/63 . So in order to integrate https://github.com/cms-sw/cmsdist/pull/9505  in to 14.1.X, do we need cmssw update too?

**Comment by hqucms** [comment]
No -- https://github.com/cms-data/RecoBTag-Combined/pull/62 only adds new models.
